woocommerce/WooCommerce-Shared
# Publish React Native Android Libraries

This PR is built on #10 and adds the capability to publish `react-native` Android libraries alongside our `library` module.

`react-native` Android libraries are made available to the `library` module through `node_modules` which is not available in WCAndroid. In Android ecosystem, these libraries should also be published to a Maven repository, however that's not the case - not for all libraries anyway. So, we have to publish them ourselves so they can be accessed by WCAndroid.

It could have been possible to only publish these libraries when they are updated in the `package.json`; which is the approach we take with `gutenberg-mobile` project using [react-native-libraries-publisher](https://github.com/wordpress-mobile/react-native-libraries-publisher). However, the isolation from the actual project makes this a tricky workflow for the developers. Furthermore, when there are changes to the publisher but not the underlying library, we have to use a [separate versioning to publish the artifacts again.](https://github.com/wordpress-mobile/react-native-libraries-publisher/blob/trunk/build.gradle.kts#L4-L18)

For this project, I wanted to try to improve the workflow especially since we have a lot less libraries to publish. So, instead of publishing these artifacts only when they are updated, we publish them whenever we publish the `library` module. To do this, we discard the actual version number from `package.json` and use the same version value the `library` module is published with. So, for example, for this PR's 8836b6722a7a66e3dcaf18b4bdecd2b3780d3f62 commit, these dependencies would like as such:

```
implementation "com.woocommerce.shared:library:57-8836b6722a7a66e3dcaf18b4bdecd2b3780d3f62"
implementation "com.woocommerce.shared:react-native-async-storage_async-storage:57-8836b6722a7a66e3dcaf18b4bdecd2b3780d3f62"
implementation "com.woocommerce.shared:react-native-safe-area-context:57-8836b6722a7a66e3dcaf18b4bdecd2b3780d3f62"
implementation "com.woocommerce.shared:react-native-screens:57-8836b6722a7a66e3dcaf18b4bdecd2b3780d3f62"
```

Since these `react-native` libraries are already dependencies of the `library` module, they don't need to be added separately. So, this is the only dependency needed in WCAndroid:

```
implementation "com.woocommerce.shared:library:57-8836b6722a7a66e3dcaf18b4bdecd2b3780d3f62"
```

---

In order to get this setup working, we need to first publish the `react-native` libraries. This capability is added to the root `build.gradle` file [as such](https://github.com/woocommerce/WooCommerce-Shared/pull/57/files#diff-6374e868882020bf62bf89806fdabed8f0edb27ed6d7ab8b9901e33cd40e5749R48-R81). This configures the projects defined in `ext.reactNativeSubProjectsToPublish` to add them the necessary plugins and the `publishing` block. Once this is added, we are able to publish each of these projects from Buildkite [like so](https://github.com/woocommerce/WooCommerce-Shared/pull/57/files#diff-cf3599ea289bab6462019b62fcb38d657f2e004802254b43c4b4ebf62aec0ff5R26-R40). It's important for `REACT_NATIVE_PROJECTS_TO_PUBLISH` variable in the bash script and `reactNativeSubProjectsToPublish` variable in the root `build.gradle` file to match, which is why they have cross comments referencing each other.

After the `react-native` libraries are published, we are still not able to publish the `library` module yet. Or rather, if we do that, we'll end up with a POM file where the version of these `react-native` library dependencies would still be missing. This is because the project by default is configured to use the dependencies in the `node_modules` folder. However, during publishing we need to use the dependencies that we just published to our S3 Maven. The way to handle this in Gradle is to use[ dependency substitution](https://github.com/woocommerce/WooCommerce-Shared/pull/57/files#diff-149900d4c8d30f55a0bd21b534ede8486a578d3a6a80bee3f40c3d335047ecf5R68-R78).

We are almost ready to publish the `library` module but there is still one more issue to solve - which should be the motto for `react-native` 🙈 The `react-native` & `hermes` dependencies are added to the project without a defined version. The reason [we do this in our library](https://github.com/woocommerce/WooCommerce-Shared/blob/8836b6722a7a66e3dcaf18b4bdecd2b3780d3f62/libraries/android/library/build.gradle#L56-L57) is because we let React Native Gradle Plugin to set the version for us - which reads it from the `package.json` file. _We actually don't need to do this anymore, because [we are also parsing this value](https://github.com/woocommerce/WooCommerce-Shared/blob/8836b6722a7a66e3dcaf18b4bdecd2b3780d3f62/libraries/android/build.gradle#L25)._ The reason it's done in most `react-native` libraries is because they use the `react-native` version from the `node_modules` folder, so they set it to `+` which means the latest version. This is not something we should allow because it'll override the version defined by us.

In order to address the versioning issue with `react-native` & `hermes`, [we tell Gradle to put the _resolved_ version of its dependencies to the POM file](https://github.com/woocommerce/WooCommerce-Shared/pull/57/files#diff-6374e868882020bf62bf89806fdabed8f0edb27ed6d7ab8b9901e33cd40e5749R71-R75) instead of the declared ones - which is the default behavior.

Sorry, there is just one more thing. The `react-native` dependency has been renamed to `react-android` and we don't want to have both of these dependencies because it'd create a duplicate class issue. So, we use dependency substitution to [address this](https://github.com/woocommerce/WooCommerce-Shared/pull/57/files#diff-6374e868882020bf62bf89806fdabed8f0edb27ed6d7ab8b9901e33cd40e5749R33-R46) which should also partially address the previous issue 🤔 

With all that in place, we are now able to publish the `library` module from Buildkite [as such](https://github.com/woocommerce/WooCommerce-Shared/pull/57/files#diff-cf3599ea289bab6462019b62fcb38d657f2e004802254b43c4b4ebf62aec0ff5R48). We do need to pass the version name of the `react-native` libraries, which is not ideal, but `publish-to-s3-gradle-plugin` [has a task](https://github.com/woocommerce/WooCommerce-Shared/pull/57/files#diff-cf3599ea289bab6462019b62fcb38d657f2e004802254b43c4b4ebf62aec0ff5R44) to calculate the same version value these projects are published with, [as long as the parameters are the same.](https://github.com/woocommerce/WooCommerce-Shared/pull/57/files#diff-cf3599ea289bab6462019b62fcb38d657f2e004802254b43c4b4ebf62aec0ff5R33) So, this is more of an annoyance than a proper issue.

**To Test**

@wzieba confirmed that this is working as expected in https://github.com/woocommerce/woocommerce-android/pull/9225. I also confirmed from the build scan that it has correct dependencies. So, I think testing can be skipped. However, referred WCAndroid branch is available for testing for the curious.
